I1354 modularization of configuring workspace exposure strategies (#853)
codenvy/che-docs: src/main/pages/che-7/end-user-guide/con_security-considerations.adoc

// configuring-workspace-exposure-strategies

[id="security-considerations_{context}"]
= Security considerations

All the security-related considerations below are only applicable to Che in multi-user mode. The single user mode does not impose any security restrictions.

> == JSON web token proxy

Reviewer: ```suggestion
[id="json-web-token-jwt-proxy_{context}"]
== JSON web token (JWT) proxy

```

> All the Che plug-ins, editors, and components can require authentication of the user accessing them. This authentication can be performed using a JSON web token (JWT) proxy that functions as a reverse proxy of the respective component based on their configuration and does the authentication on the components’ behalf. This authentication is done using a redirect to a special page on the Che server that can propagate the workspace and user-specific authentication token (workspace access token) back to the originally requested page.

Reviewer: ```suggestion
All Che plug-ins, editors, and components can require authentication of the user accessing them. This authentication is performed using a JSON web token (JWT) proxy that functions as a reverse proxy of the respective component based on its configuration and performs the authentication on behalf of the component.

The authentication uses a redirect to a special page on the Che server that propagates the workspace and user-specific authentication token (workspace access token) back to the originally requested page.
```

The JWT proxy accepts the workspace access token from the following places in the incoming requests, in the following order:

. The token query parameter
. The Authorization header in the bearer-token format

> . The access_token cookie

Reviewer: ```suggestion
. The `access_token` cookie
```

> == Secured plug-ins and editors

Reviewer: ```suggestion
== Secured plug-ins and editors

```

> Generally speaking, the end user of Che does not need to secure the workspace plug-ins and workspace editors, like for example Theia. This is because JWT proxy authentication should be transparent to the user, and is governed by the plug-in or editor definition in its `meta.yaml` descriptor.

Reviewer: ```suggestion
Che users do not need to secure workspace plug-ins and workspace editors (such as Che-Theia). This is because the JWT proxy authentication is transparent to the user and is governed by the plug-in or editor definition in their `meta.yaml` descriptors.
```

> == Secured container-image components

Reviewer: ```suggestion
== Secured container-image components

```

> The container image components can define custom endpoints for which the devfile author can require Che-provided authentication, if needed. This authentication is configured using two optional attributes of the endpoint:

Reviewer: ```suggestion
Container-image components can define custom endpoints for which the devfile author can require Che-provided authentication, if needed. This authentication is configured using two optional attributes of the endpoint:
```

> * secure - A boolean attribute  that instructs the Che server to put the JWT proxy in front of the endpoint. Such endpoints will have to be provided the workspace access token in one of the several ways explained in the mentioned above dedicated JWT Proxy chapter. This attribute value defaults to _false_.

Reviewer: ```suggestion
* `secure` - A boolean attribute that instructs the Che server to put the JWT proxy in front of the endpoint. Such endpoints have to be provided with the workspace access token in one of the several ways explained in xref:#json-web-token-jwt-proxy_{context}[]. The default value of the attribute is `false`.
```

> * cookiesAuthEnabled - A boolean attribute  that instructs the Che server to automatically redirect the unauthenticated requests for current user authentication as described in the JWT Proxy chapter above. Note that enabling this has security consequences by opening up the possibility of Cross-site request forgery (CSRF) attacks. Therefore, the value of this attribute defaults to _false_.

Reviewer: ```suggestion
* `cookiesAuthEnabled` - A boolean attribute that instructs the Che server to automatically redirect the unauthenticated requests for current user authentication as described in xref:#json-web-token-jwt-proxy_{context}[]. Setting this attribute to `true` has security consequences because it makes Cross-site request forgery (CSRF) attacks possible. The default value of the attribute is `false`.
```

> == Cross-site request forgery attacks

Reviewer: ```suggestion
== Cross-site request forgery attacks

```

> The cookie-based authentication can make the application secured by the JWT proxy prone to Cross-site request forgery (CSRF) attacks. Please read the link:https://en.wikipedia.org/wiki/Cross-site_request_forgery[Cross-site request forgery] Wikipedia page and other resources to make sure your application is not vulnerable.

Reviewer: ```suggestion
Cookie-based authentication can make an application secured by a JWT proxy prone to Cross-site request forgery (CSRF) attacks. See the link:https://en.wikipedia.org/wiki/Cross-site_request_forgery[Cross-site request forgery] Wikipedia page and other resources to ensure your application is not vulnerable.
```

> == Phishing attacks

Reviewer: ```suggestion
== Phishing attacks

```

> If an attacker is able to create an ingress or route inside the cluster with the workspace that shares the host with some of the services behind a JWT proxy, the attacker may be able to create a service and a specially forged Ingress. If such a service or Ingress is accessed by a legitimate user that was previously authenticated with a workspace, it can lead to the attacker stealing the workspace access token from the cookies sent by the legitimate user’s browser to the forged URL.

Reviewer: ```suggestion
An attacker who is able to create an Ingress or route inside the cluster with the workspace that shares the host with some of the services behind a JWT proxy, the attacker may be able to create a service and a specially forged Ingress object. When such a service or Ingress is accessed by a legitimate user that was previously authenticated with a workspace, it can lead to the attacker stealing the workspace access token from the cookies sent by the legitimate user’s browser to the forged URL.
```

> To mitigate this possibility, OpenShift can be configured to disallow setting the host of an Ingress and thus eliminating this attack vector completely.

Reviewer: ```suggestion
To eliminate this attack vector, configure OpenShift to disallow setting the host of an Ingress.
```
